src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -274,60 +274,80 @@ bool AbstractLQPNode::has_matching_unique_constraint(const ExpressionUnorderedSe
+    auto dependents = ExpressionUnorderedSet();
+    for (const auto& output_expression : output_expressions) {
+      if (!determinants.contains(output_expression)) {
+        dependents.insert(output_expression);
+      }
+    }
+
+    // Add the FD to the output if not already existing
+    if (!dependents.empty() &&
+        std::find_if(fds_in.cbegin(), fds_in.cend(), [&determinants, &dependents](const auto& fd) {
+          return (fd.determinants == determinants) && (fd.dependents == dependents);
+        }) == fds_in.cend()) {
+      fds_in.emplace_back(determinants, dependents);
     }
   }
 
-  auto& fds_in = fds_left;
-  fds_in.insert(fds_in.end(), fds_right.cbegin(), fds_right.cend());
-
-  // Currently, we do not support FDs in conjunction with null values in their determinant set.
-  // Some LQP nodes, however, change column nullability (like for instance outer joins). Therefore, we check input FDs
-  // for compliance and discard them, if necessary.
+  /**
+   * (3) Remove invalid FDs:
+   *     Currently, we do not support FDs in conjunction with null values in their determinant set. Therefore, we
+   *     have to check all FDs for compliance before returning them.
+   *     The columns of the FD's dependents set are allowed to be nullable.
+   */
   auto fds_out = std::vector<FunctionalDependency>();
 
-  // Collect non-nullable columns
-  const auto expressions = output_expressions();
+  // For the following check, we need to collect all non-nullable columns
   auto output_expressions_non_nullable = ExpressionUnorderedSet{};
-  for (auto column_id = ColumnID{0}; column_id < expressions.size(); ++column_id) {
+  for (auto column_id = ColumnID{0}; column_id < output_expressions.size(); ++column_id) {
     if (!is_column_nullable(column_id)) {
-      output_expressions_non_nullable.insert(expressions.at(column_id));
+      output_expressions_non_nullable.insert(output_expressions.at(column_id));
     }
   }
-
   for (const auto& fd : fds_in) {
     // Check whether the determinants are a subset of the current node's non-nullable output expressions
     if (std::any_of(fd.determinants.cbegin(), fd.determinants.cend(),
                     [&output_expressions_non_nullable](const auto& expression) {
                       return !output_expressions_non_nullable.contains(expression);
                     })) {
+      // Discard the FD
       continue;
     }
 
-    // We do not check the columns of the FD's dependents set since they are allowed to be nullable.
-
-    // FD remains valid, so add it to the output vector
+    // Add FD to the output
     fds_out.push_back(fd);
   }
 
   return fds_out;
 }
 
+std::vector<FunctionalDependency> AbstractLQPNode::pass_functional_dependencies() const {
+  if (left_input()) {
+    Assert(!right_input(), "Expected single input node for implicit FD forwarding. Please override this function.");
+    return left_input()->pass_functional_dependencies();
+  } else {
+    return {};
```

Comment:
When is this called? Should we `Assert(left_input && !right_input)`?

Reply:
For example, in StoredTableNode, MockNode and StaticTableNode.

Reply:
Hmm, I guess overriding the method over there does not really help with anything.